sound/soc/sof/ipc4-topology.c

```diff
@@ -231,6 +231,31 @@ static int sof_ipc4_setup_process_init_byte_kcontrol(struct snd_sof_widget *swid
 	return 0;
 }
 
+static const struct sof_ipc4_pin_format
+*sof_ipc4_get_loadable_module_pin_fmt(struct snd_sof_widget *swidget, bool pin_type, int pin_id)
+{
+	struct sof_ipc4_process *process = swidget->private;
+	struct sof_ipc4_control_data *control_data = process->init_byte_kcontrol;
+	struct sof_ipc4_base_module_cfg_ext *base_ext;
+	const struct sof_ipc4_pin_format *pin_fmt;
+	int i;
+
+	if (!control_data)
+		return NULL;
+
+	base_ext = (struct sof_ipc4_base_module_cfg_ext *)control_data->data->data;
+
+	pin_fmt = (struct sof_ipc4_pin_format *)base_ext->data;
+	if (pin_type == SOF_PIN_TYPE_SINK) {
+		for (i = 0; i < base_ext->num_sink_pin_fmts; i++) {
+			if (pin_fmt[i].pin_index == pin_id)
```

Comment:
Why we treat the loadable modules so differently? Built in modules have pins, loadable modules have pins, both have source and both have sink pins, the tplg tells which pin is connected to where, so why don't we set things up that the code does not need to know if it is loadable or not, it should not matter.

Reply:
Only review comments for the last commit is not addressed now, I think more discussion are needed, I will add more background and information.

### hints on the use of term sink and source
The last commit just extend the feature https://github.com/thesofproject/linux/pull/4006, in the original PR, we are setting the additional copier output pin format from the connected input pin format of the sink widget. But the term used there is `set copier sink format`.  

The term sink/source are used differently in the firmware and the kernel. In the firmware, there is the queue buffer in between two components, say `comp1->buffer->comp2`, set_sink_format for comp1 is to set the format for the sink queue buffer. But in the kernel, there is no queue buffer concept, instead, we use sink/source pin for the input/output of a component. setting widget source(output) pin format in the kernel is the same thing as setting widget sink(the buffer) format in the firmware, and #4006 aligned with firmware.
```
/* THIS STRUCTURE DO EXIST IN BOTH SOF and REFERENCE FW */
struct sof_ipc4_copier_config_set_sink_format {
/* Id of sink */
	u32 sink_id;     // THIS IS ACTUALLY THE OUTPUT PIN ID
/*
 * Input format used by the source
 * attention must be the same as present if already initialized.
 */
	struct sof_ipc4_audio_format source_fmt;  //  THIS SHOULD BE THE COPIER INPUT PIN FORMAT 
/* Output format used by the sink */
	struct sof_ipc4_audio_format sink_fmt;    // THIS IS THE FORMAT FOR OUTPUT PIN with ID sink_id
} __packed __aligned(4);
```
@ranj063 @ujfalusi  Should we have a fixup for #4006?

Link:
- copier_set_sink_fmt call back in SOF: https://github.com/thesofproject/sof/blob/10e4e3f86bc2597a8e91da4cca40a70e88e57c82/src/audio/copier/copier.c#L1314

### why not sof_ipc4_get_sink_pin_fmt()
@ranj063 
some modules may change the output format, e.g. mic_sel widget changes channels, src changes sample rate, @libinyang has a PR to propagate the changed format to downstream by updating pipeline_params: https://github.com/thesofproject/linux/pull/4097.

As no one can guarantee that loadable modules will never change format, I think we also need to do this for loadable modules just like #4097 did. I think we also need the loadable output format later. So I write a function to support to get any pin format for loadable module. This function is not applicable for basefw module, because of the optional output format.

Reply:
It all depends _what_ `struct sof_ipc4_copier_config_set_sink_format` is applying to.
**A)** If it is applying to the connection between two modules then:
source_fmt: the output of the source module
sink_fmt: the input of the destination module

**B)** If it applies to the module itself then:
source_fmt: the input of the module
sink_fmt: the output of the module

Which one is it? I would guess that it is the **A)** But based on @aiChaoSONG's comment and annotated struct it is probably not.
In any case, we must document it clearly and then we can align the naming and the code around to be less confusing.

Reply:
> This means the stucture definitions in https://github.com/thesofproject/linux/commit/25b7d448672b64eb385ed9602d736f38b7ed8ada will also have to change accordingly.

And maybe also change the code for sink/source pin binding and copier sink setting in #3757 and #4006 accordingly?

Of course we can add comments for the code in #4006 to clarify the terms used in firmware and kernel are different. I think this is the only place that brings the confusion, others are good.

Reply:
this thread is impossible to follow at this point....